xgcm/grid_ufunc.py

```diff
@@ -593,8 +612,36 @@ def apply_as_grid_ufunc(
     if grid is None:
         raise ValueError("Must provide a grid object to describe the Axes")
 
-    if any(not isinstance(arg, xr.DataArray) for arg in args):
-        raise TypeError("All data arguments must be of type DataArray")
+    # TODO: we need some more comprehensive tests that check proper padding with
+    # TODO: e.g. multiple vector components or mixed (tracer, vector components)
+    if not all(isinstance(a, xr.DataArray) or isinstance(a, dict) for a in args):
```

Comment:
```suggestion
    if not all(isinstance(a, (xr.DataArray, dict)) for a in args):
```
You can check against multiple types by using a tuple, like this.

But also this isn't really a sufficient check, because it doesn't check that the dicts only contain 1 dataarray each. It's important to check this here because it's a place where it's likely that users get confused and try inputting the wrong types.

Reply:
Oh neat! I didnt actually know that. I think we need to first add some tests for the cases where we want errors to be raised and then refactor the input checking as you suggested.